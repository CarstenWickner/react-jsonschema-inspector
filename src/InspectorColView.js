import PropTypes from "prop-types";
import React, { PureComponent } from "react";

import InspectorColumn from "./InspectorColumn";
import JsonSchemaPropType from "./JsonSchemaPropType";

class InspectorColView extends PureComponent {
    componentDidUpdate(prevProps) {
        const previousColumnCount = prevProps.columnData.length;
        // eslint-disable-next-line react/destructuring-assignment
        const currentColumnCount = this.props.columnData.length;
        if (previousColumnCount < currentColumnCount) {
            // auto-scroll to the far right if an additional column was added
            this.colViewContainerRef.scrollLeft = this.colViewContainerRef.scrollWidth;
        }
    }

    render() {
        const {
            columnData, refTargets, appendEmptyColumn, renderItemContent
        } = this.props;
        return (
            <div className="jsonschema-inspector-colview" ref={(ref) => { this.colViewContainerRef = ref; }}>
                {columnData.map((singleColumnData, index) => (
                    <InspectorColumn
                        // eslint-disable-next-line react/no-array-index-key
                        key={index}
                        refTargets={refTargets}
                        renderItemContent={renderItemContent}
                        {...singleColumnData}
                    />
                ))}
                {appendEmptyColumn
                    && <div className="jsonschema-inspector-column-placeholder" />}
            </div>
        );
    }
}

InspectorColView.propTypes = {
    columnData: PropTypes.arrayOf(PropTypes.shape({
        items: PropTypes.objectOf(JsonSchemaPropType).isRequired,
        selectedItem: PropTypes.string,
        trailingSelection: PropTypes.bool
    })).isRequired,
    refTargets: PropTypes.objectOf(JsonSchemaPropType).isRequired,
    appendEmptyColumn: PropTypes.bool,
    renderItemContent: PropTypes.func // func({ string: name, boolean: hasNestedItems, boolean: selected, JsonSchema: schema, refTargets })
};

InspectorColView.defaultProps = {
    appendEmptyColumn: false,
    renderItemContent: null
};

export default InspectorColView;
