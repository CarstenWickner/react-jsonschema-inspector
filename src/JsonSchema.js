import RefScope from "./RefScope";
import {
    isNonEmptyObject, listValues, mapObjectValues, mergeObjects
} from "./utils";

class JsonSchema {
    /**
     * Alternative factory method to the constructor, returning "undefined" if the given schema is not a non-empty object.
     *
     * @param {Object} rawSchema the JSON Schema to represent
     * @param {Object} parserConfig configuration affecting how the JSON schema is being traversed/parsed
     * @param {RefScope} scope collection of available $ref targets
     * @returns {JsonSchema|undefined}
     */
    static createIfNotEmpty(rawSchema, parserConfig, scope) {
        return isNonEmptyObject(rawSchema) ? new JsonSchema(rawSchema, parserConfig, scope) : undefined;
    }

    /**
     * Raw JSON Schema
     */
    schema;

    /**
     * Configuration steering how the JSON schema is being traversed/parsed
     */
    parserConfig;

    /**
     * @type RefScope
     */
    scope;

    /**
     * Constructor for a JsonSchema (wrapper).
     *
     * @param {Object} schema the JSON Schema to represent
     * @param {Object} parserConfig configuration affecting how the JSON schema is being traversed/parsed
     * @param {?RefScope} scope collection of available $ref targets
     */
    constructor(schema, parserConfig, scope) {
        this.schema = schema;
        this.parserConfig = parserConfig;
        this.scope = scope || new RefScope(this);
    }

    getRelevantSchemaParts() {
        const { allOf, anyOf, oneOf } = this.schema;
        const includeAnyOf = anyOf && this.parserConfig && this.parserConfig.anyOf === "likeAllOf";
        const includeOneOf = oneOf && this.parserConfig && this.parserConfig.oneOf === "likeAllOf";
        if (allOf || includeAnyOf || includeOneOf) {
            let result = allOf;
            if (includeAnyOf) {
                result = result ? result.concat(anyOf) : anyOf;
            }
            if (includeOneOf) {
                result = result ? result.concat(oneOf) : oneOf;
            }
            return result;
        }
        return undefined;
    }

    /**
     * Look-up of values associated with a specific kind of field in the given JSON schema definition.
     * If the targeted schema is an array, the sub-schemas of its "items" will be ignored.
     *
     * @param {String} fieldName name/key of the field to look-up
     * @param {?Function} mergeFunction optional: how to combine two encountered values into one
     * @param {?Function} mappingFunction optional: converting the raw value before calling the mergeFunction
     * @return {*} result of the given mergeFunction over all encountered values of fields with the given name
     */
    getFieldValue(fieldName, mergeFunction = listValues, mappingFunction) {
        if (!isNonEmptyObject(this.schema)) {
            return undefined;
        }
        if (this.schema.$ref) {
            // schema variable is merely a reference to a separately defined schema
            const referencedSchema = this.scope.find(this.schema.$ref);
            // by convention, if $ref is specified, all other properties in the schema are being ignored
            return referencedSchema.getFieldValue(fieldName, mergeFunction, mappingFunction);
        }
        const rawValue = this.schema[fieldName];
        let value = mappingFunction ? mappingFunction(rawValue, this.parserConfig, this.scope) : rawValue;
        const schemaParts = this.getRelevantSchemaParts();
        if (schemaParts) {
            // schema variable is supposed to be combined with a given list of sub-schemas
            value = schemaParts
                .map(part => new JsonSchema(part, this.parserConfig, this.scope).getFieldValue(fieldName, mergeFunction, mappingFunction))
                .reduce(mergeFunction, value);
        }
        return value;
    }

    /**
     * Version of the getFieldValue() function to look-up of nested JSON schema definitions (while preserving the appropriate reference scope).
     *
     * This method is intended to be used to look-up fields like "items" or "additionalItems".
     *
     * @param {String} fieldName name/key of the field to look-up
     * @param {Function} mergeFunction optional: how to combine two encountered values into one
     * @return {*} result of the given mergeFunction over all encountered values of fields with the given name
     */
    getSchemaFieldValue(fieldName, mergeFunction = listValues) {
        return this.getFieldValue(fieldName, mergeFunction, JsonSchema.createIfNotEmpty);
    }

    /**
     * Determine whether this schema represents an array and if so, return the sub-schema describing its contents.
     *
     * @returns {JsonSchema} type of array contents or null if it is not an array or its content type could not be determined
     */
    getTypeOfArrayItems() {
        return this.getSchemaFieldValue("items") || this.getSchemaFieldValue("additionalItems") || null;
    }

    /**
     * Look-up of any nested sub-schemas that may contain information about "properties" in the given JSON schema definition.
     * If the given JSON schema definition refers is an array, it is assumed to not have relevant information.
     * Instead, the referenced schemas of its "Items" will be returned.
     *
     * This method is intended to be used to look-up fields like "properties" or "required" which are not associated with arrays.
     *
     * BEWARE: if a schema consists of multiple sub-schemas and only one of them defines an "items" field, the other sibling schemas
     * will still be included in the result list - they are just assumed to not have any relevant fields we are looking for.
     *
     * @returns {Array<JsonSchema>} array listing sub-schemas that may define properties about their children
     */
    getPropertyParentSchemas() {
        if (!isNonEmptyObject(this.schema)) {
            return [];
        }
        if (this.schema.$ref) {
            const referencedSchema = this.scope.find(this.schema.$ref);
            // this schema is just a reference to another separately defined schema
            return referencedSchema.getPropertyParentSchemas();
        }
        const schemaParts = this.getRelevantSchemaParts();
        if (schemaParts) {
            let returnedSchemas = [this];
            schemaParts.forEach((rawSchemaPart) => {
                returnedSchemas = returnedSchemas.concat(new JsonSchema(rawSchemaPart, this.parserConfig, this.scope).getPropertyParentSchemas());
            });
            return returnedSchemas;
        }
        if (isNonEmptyObject(this.schema.items)) {
            // unsupported: specifying array of schemas referring to entries at respective positions in described array
            // schema.items contains a single schema, that specifies the type of any value in the described array
            return new JsonSchema(this.schema.items, this.parserConfig, this.scope).getPropertyParentSchemas();
        }
        if (isNonEmptyObject(this.schema.additionalItems)) {
            // the given schema is an array which may specify its content type in "additionalItems"
            return new JsonSchema(this.schema.additionalItems, this.parserConfig, this.scope).getPropertyParentSchemas();
        }
        // if there are no nested schemas, only this schema itself may contain properties
        return [this];
    }

    /**
     * Extract the properties mentioned in this schema â€“ also considering all nested sub-schema by using getPropertyParentSchemas().
     *
     * @returns {Object.<String, JsonSchema>} collection of all properties mentioned in this schema
     */
    getProperties() {
        return this.getPropertyParentSchemas().map(({ schema, scope }) => {
            // properties is an Object.<String, raw-json-schema>
            const { properties, required = [] } = schema;
            const rawProperties = Object.assign(
                {},
                ...required.map(value => ({ [value]: true })),
                isNonEmptyObject(properties) ? properties : {}
            );
            return mapObjectValues(rawProperties, rawSchema => new JsonSchema(rawSchema, this.parserConfig, scope));
        }).reduce(mergeObjects, {});
    }
}

export default JsonSchema;
