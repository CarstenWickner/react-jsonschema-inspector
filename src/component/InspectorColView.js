import PropTypes from "prop-types";
import React, { Component } from "react";

import InspectorColumn from "./InspectorColumn";
import InspectorOptionsColumn from "./InspectorOptionsColumn";
import { getColumnDataPropTypeShape } from "./renderDataUtils";

class InspectorColView extends Component {
    componentDidUpdate(prevProps) {
        const previousColumnCount = prevProps.columnData.length + (prevProps.appendEmptyColumn ? 1 : 0);
        const { columnData, appendEmptyColumn } = this.props;
        const currentColumnCount = columnData.length + (appendEmptyColumn ? 1 : 0);
        if (previousColumnCount < currentColumnCount) {
            // auto-scroll to the far right if an additional column was added
            this.colViewContainerRef.scrollLeft = this.colViewContainerRef.scrollWidth;
        }
    }

    render() {
        const {
            columnData, appendEmptyColumn, renderItemContent
        } = this.props;
        return (
            <div
                className="jsonschema-inspector-colview"
                ref={(ref) => { this.colViewContainerRef = ref; }}
                tabIndex={-1}
            >
                {columnData.map((singleColumnData, index) => {
                    const {
                        items, options, contextGroup, selectedItem, trailingSelection, filteredItems, onSelect
                    } = singleColumnData;
                    if (items) {
                        return (
                            <InspectorColumn
                                // eslint-disable-next-line react/no-array-index-key
                                key={index}
                                items={items}
                                selectedItem={selectedItem}
                                trailingSelection={trailingSelection}
                                filteredItems={filteredItems}
                                onSelect={onSelect}
                                renderItemContent={renderItemContent}
                            />
                        );
                    }
                    return (
                        <InspectorOptionsColumn
                            // eslint-disable-next-line react/no-array-index-key
                            key={index}
                            options={options}
                            contextGroup={contextGroup}
                            selectedItem={selectedItem}
                            trailingSelection={trailingSelection}
                            filteredItems={filteredItems}
                            onSelect={onSelect}
                            renderItemContent={renderItemContent}
                        />
                    );
                })}
                {appendEmptyColumn
                    && <div className="jsonschema-inspector-column-placeholder" />}
            </div>
        );
    }
}

InspectorColView.propTypes = {
    columnData: PropTypes.arrayOf(PropTypes.shape(getColumnDataPropTypeShape(true))).isRequired,
    appendEmptyColumn: PropTypes.bool,
    renderItemContent: PropTypes.func // func({ string: name, boolean: hasNestedItems, boolean: selected, JsonSchema: schema })
};

InspectorColView.defaultProps = {
    appendEmptyColumn: false,
    renderItemContent: null
};

export default InspectorColView;
