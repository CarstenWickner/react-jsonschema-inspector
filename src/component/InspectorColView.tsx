import PropTypes from "prop-types";
import React, { Component } from "react";

import InspectorColumn from "./InspectorColumn";
import InspectorOptionsColumn from "./InspectorOptionsColumn";
import { getColumnDataPropTypeShape } from "./renderDataUtils";
import { RenderItemsColumn, RenderOptionsColumn, RenderItemContentFunction } from "../types/Inspector";

interface ColViewDefaultProps {
    appendEmptyColumn: boolean,
    renderItemContent: RenderItemContentFunction
};

interface ColViewProps extends ColViewDefaultProps {
    columnData: Array<RenderItemsColumn | RenderOptionsColumn>
};

class InspectorColView extends Component<ColViewProps> {
    private colViewContainerRef: React.RefObject<HTMLDivElement>;

    constructor(props: ColViewProps) {
        super(props);
        this.colViewContainerRef = React.createRef();
    }

    componentDidUpdate(prevProps: ColViewProps) {
        const previousColumnCount = prevProps.columnData.length + (prevProps.appendEmptyColumn ? 1 : 0);
        const { columnData, appendEmptyColumn } = this.props;
        const currentColumnCount = columnData.length + (appendEmptyColumn ? 1 : 0);
        if (previousColumnCount < currentColumnCount) {
            // auto-scroll to the far right if an additional column was added
            this.colViewContainerRef.current.scrollLeft = this.colViewContainerRef.current.scrollWidth;
        }
    }

    render() {
        const {
            columnData, appendEmptyColumn, renderItemContent
        } = this.props;
        return (
            <div
                className="jsonschema-inspector-colview"
                ref={this.colViewContainerRef}
                tabIndex={-1}
            >
                {columnData.map((singleColumnData, index) => {
                    const {
                        selectedItem, trailingSelection, filteredItems, onSelect
                    } = singleColumnData;
                    if ((singleColumnData as RenderItemsColumn).items) {
                        return (
                            <InspectorColumn
                                // eslint-disable-next-line react/no-array-index-key
                                key={index}
                                items={(singleColumnData as RenderItemsColumn).items}
                                selectedItem={selectedItem as string}
                                trailingSelection={trailingSelection}
                                filteredItems={filteredItems as Array<string>}
                                onSelect={onSelect}
                                renderItemContent={renderItemContent}
                            />
                        );
                    }
                    const { options, contextGroup } = singleColumnData as RenderOptionsColumn;
                    return (
                        <InspectorOptionsColumn
                            // eslint-disable-next-line react/no-array-index-key
                            key={index}
                            options={options}
                            contextGroup={contextGroup}
                            selectedItem={selectedItem as Array<number>}
                            trailingSelection={trailingSelection}
                            filteredItems={filteredItems as Array<Array<number>>}
                            onSelect={onSelect}
                            renderItemContent={renderItemContent}
                        />
                    );
                })}
                {appendEmptyColumn
                    && <div className="jsonschema-inspector-column-placeholder" />}
            </div>
        );
    }

    static propTypes = {
        columnData: PropTypes.arrayOf(PropTypes.shape(getColumnDataPropTypeShape(true))).isRequired,
        appendEmptyColumn: PropTypes.bool,
        renderItemContent: PropTypes.func // func({ string: name, boolean: hasNestedItems, boolean: selected, JsonSchema: schema })
    };
    
    static defaultProps: ColViewDefaultProps = {
        appendEmptyColumn: false,
        renderItemContent: null
    };
}

export default InspectorColView;
