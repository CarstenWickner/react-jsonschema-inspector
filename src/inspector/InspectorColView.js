import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';

import InspectorColumn from './InspectorColumn';
import JsonSchemaPropType from './JsonSchemaPropType';

class InspectorColView extends PureComponent {
    render() {
        const { columnData, renderItemContent } = this.props;
        return (
            <div className="jsonschema-inspector-colview">
                {columnData.map((singleColumnData, index) => (
                    <InspectorColumn
                        key={index}
                        renderItemContent={renderItemContent}
                        {...singleColumnData}
                    />
                ))}
            </div>
        );
    }
}

InspectorColView.propTypes = {
    columnData: PropTypes.arrayOf(PropTypes.shape({
        items: PropTypes.objectOf(JsonSchemaPropType).isRequired,
        selectedItem: PropTypes.string,
        trailingSelection: PropTypes.bool,
        refTargets: PropTypes.objectOf(JsonSchemaPropType)
    })),
    renderItemContent: PropTypes.func // func(string: name, JsonSchema: schema, boolean: selected)
}

export default InspectorColView;