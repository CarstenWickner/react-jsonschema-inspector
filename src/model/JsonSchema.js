import RefScope from "./RefScope";

/**
 * Representation of a Json Schema, offering a number of convenient functions for traversing and extracting information.
 */
export default class JsonSchema {
    /**
     * Raw JSON Schema.
     *
     * @type {object}
     */
    schema;

    /**
     * Configuration steering how the json schema is being traversed/parsed.
     *
     * @type {{oneOf: ?{groupTitle: ?string}, anyOf: ?{groupTitle: ?string}}}
     */
    parserConfig;

    /**
     * @type {RefScope}
     */
    scope;

    /**
     * Constructor for a JsonSchema (wrapper).
     *
     * @param {object} schema - the JSON Schema to represent
     * @param {object} parserConfig - configuration affecting how the json schema is being traversed/parsed
     * @param {?RefScope} scope - collection of available $ref targets (will be generated based on `schema` if not provided)
     */
    constructor(schema, parserConfig, scope) {
        this.schema = schema;
        this.parserConfig = parserConfig;
        this.scope = scope || new RefScope(this);
    }
}
