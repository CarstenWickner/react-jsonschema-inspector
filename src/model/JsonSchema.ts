import { isNonEmptyObject, deriveBaseUri } from "./utils";

import { RawJsonSchema, getValueFromRawJsonSchema } from "../types/RawJsonSchema";
import { ParserConfig } from "../types/ParserConfig";

const createRawSchemaFromBoolean = (value: boolean): RawJsonSchema => (value ? {} : { not: {} });

/**
 * Representation of a Json Schema, offering a number of convenient functions for traversing and extracting information.
 */
export class JsonSchema {
    /**
     * Raw JSON Schema.
     */
    readonly schema: RawJsonSchema;

    /**
     * Configuration steering how the json schema is being traversed/parsed.
     */
    readonly parserConfig: ParserConfig;

    /**
     * Collection of allowed references to other schemas/parts.
     */
    readonly scope: RefScope;

    /**
     * Constructor for a JsonSchema (wrapper).
     *
     * @param {boolean|RawJsonSchema} schema - the JSON Schema to represent
     * @param {ParserConfig} parserConfig - configuration affecting how the json schema is being traversed/parsed
     * @param {?RefScope} scope - collection of available $ref targets (will be generated based on `schema` if not provided)
     */
    constructor(schema: boolean | RawJsonSchema, parserConfig: ParserConfig, scope?: RefScope) {
        this.schema = typeof schema === "boolean" ? createRawSchemaFromBoolean(schema) : (schema as RawJsonSchema);
        this.parserConfig = parserConfig;
        this.scope = scope || new RefScope(this);
    }
}

/**
 * Helper class to facilitate looking-up re-usable sub-schemas via the "$ref" keyword. A single RefScope instance refers to one main schema
 * and its contained "$defs"/"definitions". Other main schemas maybe used by adding their respective RefScope instances via .addOtherRefScopes().
 */
export class RefScope {
    /**
     * Base URI for resolving references.
     */
    readonly baseUri: string | null;

    /**
     * Collection of available sub-schema to be referenced via "$ref" within the originating schema.
     */
    readonly internalRefs: Map<string, JsonSchema> = new Map();

    /**
     * Collection of available sub-schema to be referenced via "$ref" within the originating schema or from other schemas.
     */
    readonly externalRefs: Map<string, JsonSchema> = new Map();

    /**
     * Array of other scopes (e.g. in case of separate schemas being provided for that purpose).
     */
    otherScopes: Array<RefScope> = [];

    /**
     * Constructor collecting all available references to contained re-usable (sub-) schemas.
     *
     * @param {JsonSchema} schema - to collect $ref-erence-able sub-schemas from
     */
    constructor(schema?: JsonSchema) {
        this.baseUri = null;
        if (!schema || !isNonEmptyObject(schema.schema)) {
            return;
        }
        // can always self-reference via an empty fragment
        this.internalRefs.set("#", schema);
        // from JSON Schema Draft 6: "$id" replaces former "id"
        const mainAlias = getValueFromRawJsonSchema(schema.schema, "$id");
        let externalRefBase: string | null;
        if (mainAlias && !mainAlias.startsWith("#")) {
            // an absolute URI can be used both within the schema itself but also from other schemas
            const mainAliasWithFragment = mainAlias.endsWith("#") ? mainAlias : `${mainAlias}#`;
            const mainAliasWithoutFragment = mainAliasWithFragment.substring(0, mainAliasWithFragment.length - 1);
            this.externalRefs.set(mainAliasWithFragment, schema);
            this.externalRefs.set(mainAliasWithoutFragment, schema);
            // for definitions, there should always be the empty fragment between the URI and the definitions path
            externalRefBase = mainAliasWithFragment;
            this.baseUri = deriveBaseUri(mainAliasWithoutFragment);
        } else {
            // no valid alias provided
            externalRefBase = null;
        }
        // in draft 2019-09 the keyword "definitions" was renamed to "$defs"
        const definitionsKeyword = schema.schema.hasOwnProperty("$defs") ? "$defs" : "definitions";
        const definitions = getValueFromRawJsonSchema(schema.schema, definitionsKeyword);
        if (isNonEmptyObject(definitions)) {
            Object.keys(definitions).forEach((key) => {
                const definition = definitions[key];
                if (isNonEmptyObject(definition)) {
                    const subSchema: JsonSchema = new JsonSchema(definition, schema.parserConfig, this);
                    const subAlias = getValueFromRawJsonSchema(definition, "$id");
                    if (subAlias) {
                        // any alias provided within definitions will only be available as short-hand in this schema
                        this.internalRefs.set(subAlias, subSchema);
                    }
                    // from JSON Schema Draft 2019-09: "$anchor" for plain text references (that should no longer be provided via $id)
                    const anchor = getValueFromRawJsonSchema(definition, "$anchor");
                    if (anchor) {
                        // any alias provided within definitions will only be available as short-hand in this schema
                        this.internalRefs.set(`#${anchor}`, subSchema);
                    }
                    // can always reference schema in definitions by its path, starting from the empty fragment
                    this.internalRefs.set(`#/${definitionsKeyword}/${key}`, subSchema);
                    if (externalRefBase) {
                        // the convention was fulfilled and the top-level schema defined an absolute URI as its "$id"
                        // this allows referencing a schema in definitions by its path, starting from that URI
                        this.externalRefs.set(`${externalRefBase}/${definitionsKeyword}/${key}`, subSchema);
                        if (anchor) {
                            // from JSON Schema Draft 2019-09: "$anchor" can also be used in combination with base URI
                            this.externalRefs.set(`${externalRefBase}${anchor}`, subSchema);
                        }
                    }
                }
            });
        }
    }

    /**
     * Add other available scope.
     *
     * @param {RefScope} refScope - other reference scope that is available (at least its external $ref-erences)
     */
    addOtherScope(refScope: RefScope): void {
        this.otherScopes.push(refScope);
    }

    /**
     * Look-up a re-usable schema by its $ref-erence.
     *
     * @param {string} ref - the "$ref" value for which to look-up the associated (sub-)schema
     * @param {boolean} includeInternalRefs - whether the "$ref" value is from within the same main schema this RefScope belongs to
     * @returns {JsonSchema} the successfully looked-up reference (or null if no match was found)
     */
    findSchemaInThisScope(ref: string, includeInternalRefs = true): JsonSchema | undefined {
        return (includeInternalRefs && this.internalRefs.get(ref)) || this.externalRefs.get(ref);
    }

    /**
     * Look-up a re-usable schema by its $ref-erence.
     *
     * @param {string} ref - the "$ref" value for which to look-up the associated (sub-)schema
     * @returns {JsonSchema} the successfully looked-up reference
     * @throws error if no match was found
     */
    find(ref: string): JsonSchema {
        let result = this.findSchemaInThisScope(ref);
        const alternativeRef = !result && this.baseUri && !ref.startsWith("#") && `${this.baseUri}${ref}`;
        if (alternativeRef) {
            result = this.findSchemaInThisScope(alternativeRef);
        }
        if (!result) {
            this.otherScopes.some((otherScope) => {
                result = otherScope.findSchemaInThisScope(ref, false);
                if (!result && alternativeRef) {
                    result = otherScope.findSchemaInThisScope(alternativeRef, false);
                }
                return result;
            });
        }
        if (result) {
            return result;
        }
        throw new Error(`Cannot resolve $ref: "${ref}"${alternativeRef ? `/"${alternativeRef}"` : ""}`);
    }
}
