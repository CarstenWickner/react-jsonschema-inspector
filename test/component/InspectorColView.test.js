import React from "react";
import { mount, shallow } from "enzyme";

import InspectorColView from "../../src/component/InspectorColView";
import JsonSchema from "../../src/model/JsonSchema";

describe("renders correctly", () => {
    it("with minimal/default props", () => {
        const component = shallow(
            <InspectorColView
                columnData={[
                    {
                        items: {
                            "Item One": new JsonSchema(),
                            "Item Two": new JsonSchema()
                        },
                        selectedItem: "Item One",
                        onSelect: () => { }
                    },
                    {
                        items: {
                            "Item One-One": new JsonSchema(),
                            "Item One-Two": new JsonSchema()
                        },
                        selectedItem: "Item One-Two",
                        trailingSelection: true,
                        onSelect: () => { }
                    },
                    {
                        items: {
                            "Item One-Two-One": new JsonSchema(),
                            "Item One-Two-Two": new JsonSchema()
                        },
                        onSelect: () => { }
                    }
                ]}
            />
        );
        expect(component).toMatchSnapshot();
    });
    it("with empty column", () => {
        const component = shallow(
            <InspectorColView
                columnData={[
                    {
                        items: {
                            "Item One": new JsonSchema(),
                            "Item Two": new JsonSchema()
                        },
                        selectedItem: "Item Two",
                        trailingSelection: true,
                        onSelect: () => { }
                    }
                ]}
                appendEmptyColumn
            />
        );
        expect(component.exists(".jsonschema-inspector-column-placeholder")).toBe(true);
    });
});
describe("update according to prop changes", () => {
    const singleColumnData = [
        {
            items: {
                "Item One": new JsonSchema(),
                "Item Two": new JsonSchema()
            },
            onSelect: () => { }
        }
    ];
    const doubleColumnData = [
        {
            items: {
                "Item One": new JsonSchema(),
                "Item Two": new JsonSchema()
            },
            selectedItem: "Item Two",
            trailingSelection: true,
            onSelect: () => { }
        },
        {
            items: {
                "Item Two-One": new JsonSchema(),
                "Item Two-Two": new JsonSchema()
            },
            onSelect: () => { }
        }
    ];

    it("setting root selection (while no empty column was displayed)", () => {
        const component = mount(
            <InspectorColView columnData={singleColumnData} />
        );
        // simulate selection in root column
        component.setProps({
            columnData: doubleColumnData
        });
        expect(component.find("InspectorColumn")).toHaveLength(2);
    });
    it("setting root selection (while empty column was displayed)", () => {
        const component = mount(
            <InspectorColView
                columnData={singleColumnData}
                appendEmptyColumn
            />
        );
        // simulate selection in root column
        component.setProps({
            columnData: doubleColumnData
        });
        expect(component.find("InspectorColumn")).toHaveLength(2);
    });
    it("clearing root selection", () => {
        const component = mount(
            <InspectorColView columnData={doubleColumnData} />
        );
        // simulate clearing the selection again (thereby adding an empty column)
        component.setProps({
            columnData: singleColumnData,
            appendEmptyColumn: true
        });
        expect(component.find("InspectorColumn")).toHaveLength(1);
    });
});
