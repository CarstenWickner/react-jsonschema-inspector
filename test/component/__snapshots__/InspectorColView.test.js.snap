// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with minimal/default props 1`] = `
<div
  className="jsonschema-inspector-colview"
  tabIndex={-1}
>
  <InspectorColumn
    filteredItems={null}
    items={
      Object {
        "Item One": JsonSchemaGroup {
          "entries": Array [],
        },
        "Item Two": JsonSchemaGroup {
          "entries": Array [],
        },
      }
    }
    key="0"
    onSelect={[Function]}
    renderItemContent={null}
    selectedItem="Item One"
    trailingSelection={false}
  />
  <InspectorOptionsColumn
    contextGroup={
      JsonSchemaGroup {
        "entries": Array [],
      }
    }
    filteredItems={null}
    key="1"
    onSelect={[Function]}
    options={
      Object {
        "groupTitle": "one of",
        "options": Array [
          Object {},
          Object {},
        ],
      }
    }
    renderItemContent={null}
    selectedItem={
      Array [
        0,
      ]
    }
    trailingSelection={true}
  />
  <InspectorColumn
    filteredItems={null}
    items={
      Object {
        "Item One-Two-One": JsonSchemaGroup {
          "entries": Array [],
        },
        "Item One-Two-Two": JsonSchemaGroup {
          "entries": Array [],
        },
      }
    }
    key="2"
    onSelect={[Function]}
    renderItemContent={null}
    selectedItem={null}
    trailingSelection={false}
  />
</div>
`;
